File: LeetCode/src/array/ValidAnagram.java
package array;


public class ValidAnagram {

    public boolean isAnagram(String s, String t) {

        if (s.length() != t.length()){
            return false;
        }

        int[] count = new int[26];

        for (char c: s.toCharArray()) {
            count[c - 'a']++;
        }

        for (char c: t.toCharArray()) {
            count[c - 'a']--;
        }

        for (int value: count) {
            if (value != 0) {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {
        ValidAnagram solution = new ValidAnagram();

        // Test cases
        String s1 = "anagram", t1 = "nagaram";
        String s2 = "rat", t2 = "car";
        String s3 = "aacc", t3 = "ccac";

        System.out.println("Is \"" + s1 + "\" an anagram of \"" + t1 + "\"? " + solution.isAnagram(s1, t1)); // Output: true
        System.out.println("Is \"" + s2 + "\" an anagram of \"" + t2 + "\"? " + solution.isAnagram(s2, t2)); // Output: false
        System.out.println("Is \"" + s3 + "\" an anagram of \"" + t3 + "\"? " + solution.isAnagram(s3, t3)); // Output: false
    }

}

File: LeetCode_Python/hash/ValidAnagram.py
from collections import Counter


class ValidAnagram:
    def isAnagram(self, s: str, t: str) -> bool:
        count1 = Counter(s)
        count2 = Counter(t)
        return count1 == count2


# Time Complexity: O(n), where n is the length of the string s or t
# Space Complexity: O(1), since the size of the counter is constant (26 letters)

if __name__ == "__main__":
    solution = ValidAnagram()

    print("Expected Output: True")
    print("Actual Output:", solution.isAnagram("anagram", "nagaram"))

    print("Expected Output: False")
    print("Actual Output:", solution.isAnagram("rat", "car"))
